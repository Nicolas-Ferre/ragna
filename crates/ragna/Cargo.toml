[package]
name = "ragna"
description = "Library for easily creating GPU-native applications"
readme = "../../README.md"
keywords = ["gpu", "rendering", "compute", "shader"]
categories = ["game-development", "graphics", "rendering", "concurrency"]
exclude = [".github", "README.md"]
version.workspace = true
authors.workspace = true
edition.workspace = true
license.workspace = true
repository.workspace = true
rust-version.workspace = true

[dependencies]
bitfield-struct.workspace = true
derive-where.workspace = true
futures.workspace = true
fxhash.workspace = true
itertools.workspace = true
once_cell.workspace = true
wgpu.workspace = true
ragna_derive.workspace = true

[lints]
workspace = true
