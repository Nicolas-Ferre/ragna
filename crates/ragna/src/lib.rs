//! A library for creating easily a GPU-native application.

mod app;
mod operations;
mod operators;
mod runner;
mod types;
mod wgsl;

pub use app::*;
pub use operators::*;
pub use types::*;

/// Transforms a Rust module to a GPU module.
pub use ragna_derive::gpu;

// TODO: adapt compilation tests
