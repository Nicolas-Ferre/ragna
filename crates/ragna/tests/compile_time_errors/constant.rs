use ragna::{Const, Gpu, GpuContext};

fn main() {}

const CONSTANT: Gpu<i32, Const> = Gpu::constant(30);

fn run(ctx: &mut GpuContext) {
    CONSTANT.assign(ctx, Gpu::constant(20));
}
