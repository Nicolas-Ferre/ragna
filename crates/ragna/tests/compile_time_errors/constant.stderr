error[E0308]: mismatched types
 --> tests/compile_time_errors/constant.rs:8:9
  |
8 |         CONSTANT = 20;
  |         ^^^^^^^^^^^^^
  |         |
  |         expected `&mut Gpu<i32, Mut>`, found `&mut Gpu<i32, Const>`
  |         arguments to this function are incorrect
  |
  = note: expected mutable reference `&mut Gpu<i32, Mut>`
             found mutable reference `&mut Gpu<i32, Const>`
note: method defined here
 --> src/values.rs
  |
  |     pub fn assign(&mut self, ctx: &mut GpuContext, value: Gpu<T, impl Any>) {
  |            ^^^^^^
